### PR TITLE
Do not set source locations for small internal assembly routines.

### LINKED ISSUE #9006: Sourcemaps have less information after 0.6.3
## Description

Starting from 0.6.3, sourcemaps seem to have less information, at least around `revert`s statements.

## Environment

- Compiler version: 0.6.2, 0.6.3 and 0.6.8
- Target EVM version (as per compiler settings): Default
- Framework/IDE (e.g. Truffle or Remix): Plain solc
- EVM execution environment / backend / blockchain client: Not relevant
- Operating system: Ubuntu

## Steps to Reproduce

Given this `input.json`:

```solidity
{
  "language": "Solidity",
  "sources": {
    "c.sol": {
      "content": "pragma solidity ^0.6.0;\n\ncontract C {\n  function test() public {\n    revert();\n  }\n}\n"
    }
  },
  "settings": {
    "optimizer": {
      "enabled": false,
      "runs": 200
    },
    "debug": {},
    "outputSelection": {
      "*": {
        "*": [
          "abi",
          "evm.bytecode",
          "evm.deployedBytecode",
          "evm.methodIdentifiers"
        ],
        "": [
          "id",
          "ast"
        ]
      }
    }
  }
}
```

Compiling it with 0.6.2 and getting `deployedBytecode.sourceMap` for `c.sol` gives this:

```
25:59:0:-:0;;;;8:9:-1;5:2;;;30:1;27;20:12;5:2;25:59:0;;;;;;;;;;;;;;;;;;;40:42;;;:::i;:::-;;;69:8;;
```

While compiling with 0.6.3 (or 0.6.8; the result is the same) results in this:

```
25:59:0:-:0;;;;5:9:-1;2:2;;;27:1;24;17:12;2:2;25:59:0;;;;;;;;;;;;;;;;12:1:-1;9;2:12;40:42:0;;;:::i;:::-;;;12:1:-1;9;2:12
```

Notice that there's no mapping for the `revert()` statement: in the first sourcemap that's given by `69:8`, but there's nothing like that in the second one.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/9006

After this change, we only have source references for the abi functions.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ Compiler Features:
  * SMTChecker: Support array ``length``.
  * SMTChecker: Support array ``push`` and ``pop``.
  * Add support for natspec comments on state variables.
+ * Code Generator: Do not introduce new source references for small internal routines.
```
Comment (leonardoalt) [inline Changelog.md]: Shouldn't we try to keep the Changelog sorted already in the PRs?

File: libyul/AsmParser.h [hunk 2]
```diff
@@ -91,6 +103,7 @@ class Parser: public langutil::ParserBase
 
 private:
 	Dialect const& m_dialect;
+	std::optional<langutil::SourceLocation> m_locationOverride;
```
Comment (leonardoalt) [inline libyul/AsmParser.h:106]: Maybe just call it `m_location` and add a comment explaining why it has to override the other one?

Comment (chriseth) [reply libyul/AsmParser.h:106]: I think this is a very exceptional situation and should warrant for a long variable name.

**Review by axic — APPROVED** [review]
On this example test case it looks good, wondering in other cases, but the idea makes sense.

**Pull request merged**